concent_api/core/payments/storage.py

```diff
@@ -0,0 +1,145 @@
+
+        if global_transaction_state.nonce != tx.nonce:
+            raise Exception(
+                f'Transaction nonce does not match. '
+                f'Current={global_transaction_state.nonce}, '
+                f'Transaction={tx.nonce}'
+            )
+
+        pending_ethereum_transaction = PendingEthereumTransaction(
+            nonce       = tx.nonce,
+            gasprice    = tx.gasprice,
+            startgas    = tx.startgas,
+            value       = tx.value,
+            v           = tx.v,
+            r           = tx.r,
+            s           = tx.s,
+            data        = tx.data,
+            to          = tx.to,
+        )
+        pending_ethereum_transaction.full_clean()
+        pending_ethereum_transaction.save()
+
+        global_transaction_state.nonce += 1
+        global_transaction_state.full_clean()
+        global_transaction_state.save()
+
+    @transaction.atomic(using='control')
+    def remove_tx(self, nonce: int) -> None:
+        """
+        Remove the transaction after it's been confirmed and doesn't have
+        to be tracked anymore.
+        """
+
+        # This is called here only to lock GlobalTransactionState
+        self._get_locked_global_transaction_state()
+
+        try:
+            pending_ethereum_transaction = PendingEthereumTransaction.objects.get(nonce=nonce)
+            pending_ethereum_transaction.delete()
+            logger.info(f'Successfully removed PendingEthereumTransaction with nonce {nonce}.')
+        except PendingEthereumTransaction.DoesNotExist:
+            logger.error(f'Trying to remove PendingEthereumTransaction with nonce {nonce} but it does not exist.')
+            raise
+
+    @transaction.atomic(using='control')
+    def revert_last_tx(self) -> None:
+        """
+        Remove the last transaction that was added.
+        This shouldn't be ever called if everything is being used correctly,
+        i.e. we don't try to send invalid transactions.
+        """
+        global_transaction_state = self._get_locked_global_transaction_state()
+
+        try:
+            pending_ethereum_transaction = PendingEthereumTransaction.objects.get(
+                nonce=global_transaction_state.nonce - 1
+            )
+            pending_ethereum_transaction.delete()
+            logger.info(
+                f'Successfully reverted last PendingEthereumTransaction with nonce {pending_ethereum_transaction.nonce}.'
+            )
+        except GlobalTransactionState.DoesNotExist:
+            logger.error(
+                f'Trying to revert last PendingEthereumTransaction with nonce {global_transaction_state.nonce - 1} but it does not exist.'
+            )
+            raise
+
+    @staticmethod
+    def _get_locked_global_transaction_state() -> GlobalTransactionState:
+        try:
+            return GlobalTransactionState.objects.select_for_update().get(
+                pk=0,
+            )
```

Comment:
You're never releasing the lock explicitly. From the docs I see that the lock lasts "until the end of the transaction block". The thing is that we will likely have more than one transaction block and they're nested: one explicitly started with `@transaction.atomic` above and one that's by default on the view (due to `ATOMIC_REQUESTS`). Could you verify that it's the former that releases the lock? We definitely don't want to keep the row locked until the end of the view.

Reply:
I have checked it positively however I don't really know how to test it negatively (to cause crash because of lock). Please consider the following code:

```
with transaction.atomic(using='control'):
    with transaction.atomic(using='control'):
        s = Subtask.objects.select_for_update().get(pk=7)
    s = Subtask.objects.get(pk=7)
```

It works just fine. Do you think its enough ?

Reply:
Like this?

``` python
with transaction.atomic(using='control'):
    with transaction.atomic(using='control'):
        s = Subtask.objects.select_for_update().get(pk=7)
    s = Subtask.objects.select_for_update().get(pk=7)
```

This should deadlock if you run it and unless the first `atomic()` releases the lock.

Reply:
This code works without errors, so the inner lock is released with the end of context manager.